File: app/src/main/java/com/averycorp/prismtask/domain/usecase/DependencyCycleGuard.kt
package com.averycorp.prismtask.domain.usecase

import com.averycorp.prismtask.data.local.entity.TaskDependencyEntity

/**
 * Detects whether adding a `(blocker, blocked)` edge would introduce a
 * cycle in the task-dependency graph.
 *
 * Mirrors the lineage + depth-cap pattern in
 * `AutomationEngine.kt:73-86`: a pure DFS over the existing edges
 * looking for a path from `blocked` back to `blocker`. If found, the
 * edge would close a cycle and must be rejected at write time.
 *
 * The walk carries a [MAX_DEPTH] safety cap so a malformed graph (e.g.
 * a pre-existing cycle that slipped past validation) cannot stall the
 * caller in an infinite loop.
 *
 * Self-edges are treated as trivial cycles.
 */
object DependencyCycleGuard {

    /** Hard ceiling on path length explored before the walk gives up. */
    const val MAX_DEPTH: Int = 10_000

    /**
     * @param edges Snapshot of every existing edge in the database.
     * @param blocker Proposed blocker task id.
     * @param blocked Proposed blocked task id.
     * @return `true` if adding `(blocker, blocked)` would close a cycle.
     */
    fun wouldCreateCycle(
        edges: List<TaskDependencyEntity>,
        blocker: Long,
        blocked: Long
    ): Boolean {
        if (blocker == blocked) return true
        if (edges.isEmpty()) return false

        // Index existing edges by their tail (blocker) so traversal from
        // a node hits all outgoing edges in O(degree) instead of O(|E|).
        val outgoing = HashMap<Long, MutableList<Long>>(edges.size)
        for (edge in edges) {
            outgoing.getOrPut(edge.blockerTaskId) { mutableListOf() }
                .add(edge.blockedTaskId)
        }

        // DFS from `blocked` along existing edges. If we ever reach
        // `blocker`, the proposed edge `(blocker, blocked)` would close
        // a loop: blocker → blocked → … → blocker.
        val visited = HashSet<Long>()
        val stack = ArrayDeque<Long>()
        stack.addLast(blocked)
        var depth = 0
        while (stack.isNotEmpty()) {
            depth++
            if (depth > MAX_DEPTH) return false
            val current = stack.removeLast()
            if (!visited.add(current)) continue
            val children = outgoing[current] ?: continue
            for (child in children) {
                if (child == blocker) return true
                stack.addLast(child)
            }
        }
        return false
    }
}

File: web/src/utils/dependencyCycleGuard.ts
/**
 * Detects whether adding a `(blocker, blocked)` edge would introduce a
 * cycle in the task-dependency graph.
 *
 * TypeScript port of Android `domain/usecase/DependencyCycleGuard.kt`
 * (DFS over outgoing edges, `MAX_DEPTH=10_000`). Used by the Roadmap
 * dependency picker to reject cycle-closing edges before write —
 * Android enforces the same guard server-side, but rejecting on web
 * gives the user immediate feedback and avoids a round-trip.
 *
 * Self-edges are treated as trivial cycles. A pre-existing cycle in
 * the input graph is bounded by [MAX_DEPTH] so the walk cannot stall.
 */

import type { TaskDependency } from '@/types/taskDependency';

export const MAX_DEPTH = 10_000;

export function wouldCreateCycle(
  edges: ReadonlyArray<Pick<TaskDependency, 'blocker_task_id' | 'blocked_task_id'>>,
  blocker: string,
  blocked: string,
): boolean {
  if (blocker === blocked) return true;
  if (edges.length === 0) return false;

  const outgoing = new Map<string, string[]>();
  for (const edge of edges) {
    const list = outgoing.get(edge.blocker_task_id);
    if (list) {
      list.push(edge.blocked_task_id);
    } else {
      outgoing.set(edge.blocker_task_id, [edge.blocked_task_id]);
    }
  }

  const visited = new Set<string>();
  const stack: string[] = [blocked];
  let depth = 0;
  while (stack.length > 0) {
    depth++;
    if (depth > MAX_DEPTH) return false;
    const current = stack.pop()!;
    if (visited.has(current)) continue;
    visited.add(current);
    const children = outgoing.get(current);
    if (!children) continue;
    for (const child of children) {
      if (child === blocker) return true;
      stack.push(child);
    }
  }
  return false;
}
